### PR TITLE
Fix retry behaviour in cluster

### PR DESCRIPTION
**Pull request by neunhoef**
The behaviour of network::sendRequestRetry is found lacking in multiple aspects:

  - the code for a follower refusing leader work should be on the list of cases to retry,
    since a former leader will respond with this and a coordinator could
    still talk to the old leader and then should move on to the new leader
    and retry, we decided to use a new code 421 Misdirected Request for this.

 -  sendRequestRetry currently does not distinguish the cases of
    "Connection refused" (nothing can have happened on the other side)
    and "Connection Closed" (potentially broken pipe during the send or
    when waiting for response, in which case the request might have gone
    through, must not retry in most cases) and "Timeout" (request could have
    gone through, must not retry in most cases) and "Connection Cancelled"
    (request was locally cancelled, should retry to allow for failover).

Lot's of `sendRequest` calls should actually be `sendRequestRetry`.

And: Make sure that we cover the case of a stale connection coming from
the connection pool and do a retry if the connection breaks on first
using it.

Furthermore, we improve the retry behavior for agency communications
in the inquire case.

Finally, a coordinator redirect (for a query cursor or so) which finds that the coordinator
to redirect to is no longer member of a cluster must return 404 instead of 503, otherwise
the client would retry.

### Scope & Purpose

*(Please describe the changes in this PR for reviewers - **mandatory**)*

- [*] :hankey: Bugfix (requires CHANGELOG entry)

#### Backports:

- [*] Backports required for: 3.8 at least

### Testing & Verification

*(Please pick either of the following options)*

- [*] This change is already covered by existing tests, such as *(please describe tests)*.
- [*] Additional integration and unit tests have been added for the retry behavior around the
       421 status code (ex-leader refuses leader operation).

### CODE REVIEW

File: tests/Network/MethodsTest.cpp [hunk 1]
```diff
@@ -185,6 +185,44 @@ TEST_F(NetworkMethodsTest, request_with_retry_after_error) {
   ASSERT_EQ(res.statusCode(), fuerte::StatusAccepted);
 }
 
+TEST_F(NetworkMethodsTest, request_with_retry_after_421) {
```
Comment (mchacki) [inline tests/Network/MethodsTest.cpp:188]: This test is good :+1:

However this PR also modifies the behavior of:
```
fuerte::Error::ConnectionClosed:
fuerte::Error::RequestTimeout:
fuerte::StatusNotAcceptable
```

which obviously are not under existing test (they are green although behavior changed) so I would expect tests for the new behavior there as well.

I would also hope for existing behavior tests, but you can argue that this is out of scope for this PR.
The new tests are not.

File: tests/js/client/shell/shell-failures-communication-cluster.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    tearDownAll: function () {
+      internal.debugClearFailAt();
+      db._drop(cn);
+    },
+
+    tearDown: function () {
+      internal.debugClearFailAt();
+    },
+
+    testRetryOnLeaderRefusal: function() {
```
Comment (mchacki) [inline tests/js/client/shell/shell-failures-communication-cluster.js:60]: This test is good.
However it is rather complicated to understand what it actually tests (see inline comment suggestions)

Also we now need a replica of this test for all our CRUD Apis if we want to test the behavior this way.

File: tests/js/client/shell/shell-failures-communication-cluster.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      let inq = db._connection.PUT_RAW("/_api/job/" + id, {});
+      print("Expecting delay with 204:", inq);
+      assertEqual(204, inq.code);
+      assertFalse(inq.error);
+      internal.wait(2.0);
+      inq = db._connection.PUT_RAW("/_api/job/" + id, {});
+      print("Expecting delay with 204:", inq);
+      assertEqual(204, inq.code);
+      assertFalse(inq.error);
+      internal.debugClearFailAt("documents::insertLeaderRefusal");
```
Comment (mchacki) [inline tests/js/client/shell/shell-failures-communication-cluster.js:77]:
```suggestion
      // Allow the leader to write
      internal.debugClearFailAt("documents::insertLeaderRefusal");
```

**Review by mchacki — CHANGES_REQUESTED** [review]
Production code looks good in most cases, please see inline comments for a few remarks..

However I am missing tons of tests on this PR:
1) I think the sendRequest() => sendRequestRetry() places are good, and do not require further tests.
2) The sendRequestRetry() now behaves very different then before. Only one of the changes got a new test, not all of them. It also seems to not be under specific tests before as no adjustments were necessary for this behavior change.
3) The Document Methods now deliver slightly different error-codes (not tested)
4) The Document API now indirectly delivers different HTTP error codes as the Document Methods behave differently. I would like to have tests on this HTTP API, only. It should be possible to directly call this API on DBServers (leader or follower) with illegal bodies. And assert the resulting HTTP Error codes. If we have those I am happy to not have specific tests on (3) as they are covered by these tests here.

File: 3rdParty/fuerte/include/fuerte/types.h [hunk 4]
```diff
@@ -53,6 +53,7 @@ StatusCode constexpr StatusMethodNotAllowed = 405;
 StatusCode constexpr StatusNotAcceptable = 406;
 StatusCode constexpr StatusConflict = 409;
 StatusCode constexpr StatusPreconditionFailed = 412;
+StatusCode constexpr StatusMisdirectedRequest = 421;
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply 3rdParty/fuerte/include/fuerte/types.h:56]: The repo says: "As Fuerte is a c++ library that is used for internal communication in ArangoDB, this repository is deprecated and fuerte will be maintained in the ArangoDB repository." So I do not think this is necessary.

File: CHANGELOG [hunk 5]
```diff
@@ -34,6 +38,7 @@ devel
 
 * Added enterprise-build-repository and oskar-build-repository to `--version`
   as `enterprise-build-repository` and `oskar-build-repository`.
+>>>>>>> origin/devel
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply CHANGELOG:41]: Done.

File: arangod/GeneralServer/RestHandler.cpp [hunk 6]
```diff
@@ -230,23 +230,32 @@ futures::Future<Result> RestHandler::forwardRequest(bool& forwarded) {
...
-                                     std::move(payload), options, std::move(headers));
+
+  // Should the coordinator be gone by now, we'll respond with 404.
+  // There is no point forwarding requests. This affects transactions, cursors, ...
+  if (server().getFeature<ClusterFeature>().clusterInfo().getServerEndpoint(serverId).empty()) {
+    generateError(rest::ResponseCode::NOT_FOUND,
+                  TRI_ERROR_CLUSTER_SERVER_UNKNOWN,
+                  std::string("cluster server ") + serverId + " unknown");
+    return Result(TRI_ERROR_CLUSTER_SERVER_UNKNOWN);
+  }
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply arangod/GeneralServer/RestHandler.cpp:254]: Yes, this can happen, but it does not matter. In such a case we would hand out 503 as before. That is OK, since it will probably let a client do a retry and we are back here. It is only necessary that we will **eventually** return 404, such that the client (in our case the testagent) does not retry indefinitely.

File: arangod/RestHandler/RestDocumentHandler.cpp [hunk 7]
```diff
@@ -102,32 +102,6 @@ void RestDocumentHandler::shutdownExecute(bool isFinalized) noexcept {
   RestVocbaseBaseHandler::shutdownExecute(isFinalized);
 }
 
-/// @brief returns the short id of the server which should handle this request
-ResultT<std::pair<std::string, bool>> RestDocumentHandler::forwardingTarget() {
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply arangod/RestHandler/RestDocumentHandler.cpp:106]: Yes, it is obsolete. The reason is that this code in the `RestDocumentHandler` does **exactly the same** as the base class `RestVocbaseBaseHandler`. So by removing the code we remove duplicate code and do not change the behaviour.

File: arangod/RestHandler/RestImportHandler.cpp [hunk 8]
```diff
@@ -868,6 +868,10 @@ Result RestImportHandler::performImport(SingleCollectionTransaction& trx,
     if (resultSlice.isArray()) {
       size_t pos = 0;
 
+      if (resultSlice.length() != babies.slice().length()) {
+        LOG_TOPIC("b9729", ERR, Logger::FIXME) << "Unequal length in import result. resultSlice: " <<
+          resultSlice.toJson() << " babies.slice()" << babies.slice().toJson();
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply arangod/RestHandler/RestImportHandler.cpp:873]: I will remove this debugging code in its entirety. This was used to hunt a bug which is in the meantime fixed in a different PR: https://github.com/arangodb/arangodb/pull/13970

File: arangod/Transaction/Methods.cpp [hunk 9]
```diff
@@ -1042,6 +1043,10 @@ Future<OperationResult> transaction::Methods::insertLocal(std::string const& cna
     auto const& followerInfo = collection->followers();
     std::string theLeader = followerInfo->getLeader();
     if (theLeader.empty()) {
+      TRI_IF_FAILURE("documents::insertLeaderRefusal") {
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply arangod/Transaction/Methods.cpp:1046]: I have now made the failure point conditional on the document actually containing a very specific attribute which is used in the test. This makes it safe for other collections.

File: arangod/Transaction/Methods.cpp [hunk 10]
```diff
@@ -1042,6 +1043,10 @@ Future<OperationResult> transaction::Methods::insertLocal(std::string const& cna
...
+        LOG_DEVEL << "Refusing leader insert!";
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply arangod/Transaction/Methods.cpp:1047]: Done.

File: tests/Network/MethodsTest.cpp (hunk 1 above)
Comment (neunhoef) [reply tests/Network/MethodsTest.cpp:188]: I will add new tests for these cases. This should not be a big problem. I think adding tests for existing behaviour is beyond the scope of this PR.

File: tests/js/client/shell/shell-failures-communication-cluster.js (hunk 2 above)
Comment (neunhoef) [reply tests/js/client/shell/shell-failures-communication-cluster.js:60]: I accept your suggestions. However, I do not think we need to replicate this test for all CRUD APIs, since we are only testing the behaviour of the NetworkFeature, and this is used in the same way for all CRUD APIs. This test does **not** test the behaviour of insert. It merely tests something which happens to be used by insert (and other APIs). Therefore I think it is good enough to run the test for insert.

File: tests/Network/MethodsTest.cpp (hunk 1 above)
Comment (neunhoef) [reply tests/Network/MethodsTest.cpp:240]: Done.

File: tests/Network/MethodsTest.cpp [hunk 11]
```diff
@@ -147,6 +147,58 @@ TEST_F(NetworkMethodsTest, request_failure) {
   ASSERT_FALSE(res.hasResponse());
 }
 
+TEST_F(NetworkMethodsTest, request_failure_on_status_not_acceptable) {
```
Comment (mchacki) [inline tests/Network/MethodsTest.cpp:150]: We need these `request_failure-*` tests on sendRequestRetry as well.
I think they should behave the same, but the code-path differs.

**Review by mchacki — CHANGES_REQUESTED** [review]
Like changes thus far, very good. :+1:

Still open for me:
1) There are some retry tests still missing (see comment)
2) We need Inter HTTP API tests still (as discussed in Slack)

**Review by mchacki — APPROVED** [review]
LGTM.
All tests included now.
